### PR TITLE
[React] Allow SFC to return null 

### PR DESCRIPTION
**Pull request by yuit**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [ ] Run `tsc` without errors.
- [ ] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: Bug: https://github.com/Microsoft/TypeScript/issues/11566
documentation: https://github.com/facebook/react/pull/5884
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

### CODE REVIEW

File: react/index.d.ts [hunk 1]
```diff
@@ -2620,7 +2620,9 @@ declare namespace React {
 
 declare global {
     namespace JSX {
-        interface Element extends React.ReactElement<any> { }
+        interface JSXElement extends React.ReactElement<any> { }
+        type Element = JSXElement | null;
+
         interface ElementClass extends React.Component<any, any> {
             render(): JSX.Element | null;
```
Comment (DanielRosenwasser) [inline react/index.d.ts]: Looks like this should just be changed to `JSX.Element` if I'm not mistaken.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
Can you add a test for this?

File: react/index.d.ts (hunk 1 above)
Comment (yuit) [inline react/index.d.ts]: Yep!

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
Thank you Yui!

**Comment by JakeGinnivan** [comment]
This causes a bunch of issues with anything which takes a `React.Element<any>` as an argument. For instance:

```
import { mount } from 'enzyme'
import MyComponent from './my-component'

mount(<MyComponent />) // Will not compile because null | Element is not assignable to React.Element<any>
```

This patch appears to completely break usage of JSX

**Comment by shiftkey** [comment]
Seconding what @JakeGinnivan found. `@types/react@15.0.5` seems to have affected any uses of `renderIntoDocument`.

Rough repro:

```ts
import * as React from 'react'
import * as TestUtils from 'react-addons-test-utils'

class App extends React.Component<IAppProps, IAppState> {
   // ...
}

TestUtils.renderIntoDocument(<App />)   
```

Output:
```
error TS2345: Argument of type 'Element' is not assignable to parameter of type 'ReactElement<any>'.
  Type 'null' is not assignable to type 'ReactElement<any>'.
```

**Comment by bbenezech** [comment]
@DanielRosenwasser can you revert this PR? It basically broke all react setups.
@yuit I think `Test the change in your own code. (Compile and run.)` is actually the most important part.

**Comment by r00ger** [comment]
enzyme's `shallow`, `mount` and `render` are broken
```ts
export function shallow<P, S>(node: ReactElement<P>, options?: ShallowRendererProps): ShallowWrapper<P, S>;
```
Obviously
```
Argument of type 'Element' is not assignable to parameter of type 'ReactElement<any>'.
  Type 'null' is not assignable to type 'ReactElement<any>'.
```

**Comment by vutran** [comment]
I'm also experience these non-assignable warnings with my jest/react-test-renderer setups.

**Comment by levrik** [comment]
I'm currently using `@types/react@15.0.4` with that piece of code to bypass the issue fixed by this PR without inventing the other problems:
```typescript
import * as React from 'react';

interface StatelessComponent<P> extends React.StatelessComponent<P> {
  (props: P & { children?: React.ReactNode }, context?: any): React.ReactElement<any> | null;
}

export default StatelessComponent;
```

**Comment by begincalendar** [comment]
@yuit has made the changes.
@RyanCavanaugh has been requested to review this, but didn't get a chance to before it was merged.
@DanielRosenwasser has reviewed and merged this PR.

The React typing changes in this PR were brought about because of this [issue](https://github.com/Microsoft/TypeScript/issues/11566), with the specifics of the change coming from @RyanCavanaugh in this [comment](https://github.com/Microsoft/TypeScript/issues/11566#issuecomment-253306155).

Handling the return of `null` from class components was easily handled in this [PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/10611).

It appears as if handling the return of `null` from stateless function components is not as simple as @yuit thought it might be (given that the added test case did not prevent the regression).
@RyanCavanaugh you should be able to attest to this; that work is also required by TypeScript to make it happen.

Regardless, as this has been merged and people above have brought to light, it has caused lots of (seemingly unintended) breaking changes.

Given that work is required in this project and in TypeScript, how are we to proceed?

**Comment by DanielRosenwasser** [comment]
Hi all, I apologize for the inconvenience this may have caused and the delay (DefinitelyTyped gets a lot of traffic and it's hard to keep track of messages). I've discussed this with @yuit and we're going to revert the PR.

**Comment by yuit** [comment]
@JakeGinnivan @shiftkey @bbenezech @r00ger @vutran @cwmoo740 @levrik @begincalendar sorry for this inconvenience of this incorrect change. I have a PR that will revert the change for now (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/14476)

**Comment by begincalendar** [comment]
@DanielRosenwasser @yuit We're only human, mistakes will be made.

I just wanted this to get the attention it needed and lay out an explanation for everyone affected.

Thank you for the work you do.

**Comment by DanielRosenwasser** [comment]
Update: We're looking into republishing - until that point, the last known good install for `@types/react` was 15.0.4.

**Comment by afc163** [comment]
Problem is resolved in `@types/react@15.0.8`, Thx! :handshake:

**Comment by tkrotoff** [comment]
TypeScript 2.3 has been released and it should include all the needed fixes for this PR to be merged again

**Pull request merged**